### DISCUSSION TITLE
[2023-09-18] Library & Engine Team Update

### DISCUSSION THREAD

**Discussion by GregoryTravis** [enso-org/enso#7820] [Weekly Demo Day]
- James: Fixes issue writing to a dry run Excel File #7763
- James: S3 Read Access, Input Stream based reading #7776
- Radek: Support for BigInteger values in Table #7715
- Radek: More robust/friendly Date Time formatter #7826 
- Greg: New Data.post for sending a payload to a Web API #7700
- Jaroslav: Support runtime checks of intersection types #7769
- Jaroslav: Avoid StackOverflow when comparing unknown foreign objects #7780
- Jaroslav: Benchmark shall loop correctly #7781
- Jaroslav: Checking ascribed expression types for existence #7796
- Jaroslav: Moving numpy check into test/Examples_Tests #7808
- Jaroslav: Cleanup code in IdExecutionInstrument #7811
- Jaroslav: Throw UnsupportedMessageException when loosing BigInteger precision #7817
- Hubert: Cleanup serialization config #7748
- Hubert: Workaround crashes during regular execution #7778
- Hubert: Improve shutdown logic of language server #7801
- Hubert: Fix log output for suggestions' tests #7803
- Hubert: Rename Decimal to Float #7807
- Hubert: Reduce extra output in compilation and tests #7809
- Hubert: Always log verbose logs to a log file #7825 
- Pavel: Implement private modules #7304
- Dmitry: expressionUpdates do not contain Method Pointer of operators #7659
- Dmitry: Log visualization expressions generated by IDE #7756
- Dmitry: Download single edition by name #7771
- Dmitry: Fix flaky project rename test #7794
- Dmitry: Send suggestions database on request #7812

**Comment by radeusgd** [comment]
A few words on new date-time formats:

Rationale: we were having weird bugs (#7461) because in Java format `yyyy` and `YYYY` mean different things (first is a year of era, the second is a special _ISO week date_ year, e.g. `Sat 1 Jan 1977` is actually `1976-W53-6` in this format). Similarly `d` and `D` have slightly different meanings. This has caused great confusion. 

We've now developed our own, simplified format for the date-time format patterns. It is mostly case insensitive (`YYYY` and `yyyy` always mean the same thing), with the only exception of `H/h` (24 vs 12 hour clock hours) and `M` (months) vs `m` (minutes).

![image](https://github.com/enso-org/enso/assets/1436948/22313cd5-3878-4061-be13-86c578d54089)

To still allow the ISO week date, we simply use a separate function to enable this format (in it, `yyyy` always means the week-based year, and regular months are unavailable because their meaning is incompatible with week-based-years):

![image](https://github.com/enso-org/enso/assets/1436948/30df0ddf-dfa4-4ad7-afeb-31620ae089e1)

Our advanced users can still use the Raw Java format if they wish to, using `from_java`:

![image](https://github.com/enso-org/enso/assets/1436948/a3bbf0b2-3b2a-4f5b-b417-523f51190cc6)

---

This is still work in progress. In the next iterations, I plan to add some validation layers to make the user experience even better and error-less. For example, when we encounter a month field between time fields (`HH:MM:ss`) or minute field between date fields (`yyyy-mm-dd`), we will accept such cases (there can be weird formats out there), BUT we will attach a warning telling the user that this is likely a mistake. Similarly, a 24h hour with an ampm marker (`HH:mm a`) which is redundant, will also yield a warning.

**Comment by jdunkerley** [comment]
S3 Integration:
![image](https://github.com/enso-org/enso/assets/4699705/6437f56a-3ab5-44c5-990a-9c6658eee62a)
![image](https://github.com/enso-org/enso/assets/4699705/4b874ba0-b48e-440e-b555-8e3ea338fd00)
